aten/src/ATen/native/cuda/DepthwiseConv2d.cu

```diff
@@ -1,29 +1,597 @@
+    conv_depthwise2d_grad_weight_kernel<<<grid, block, smem, stream>>>(
+        grad_output_a, input_a, grad_weight_a, batchSize, inputChannels, outputChannels, depthwiseMultiplier,
+        width, height, outputWidth, outputHeight, kW, kH, dW, dH, padW, padH, dilationW, dilationH);
+    C10_CUDA_KERNEL_LAUNCH_CHECK();
+  });
+}
+
+}  // namespace (anonymous)
+
+const Tensor& conv_depthwise2d_cuda_out(
+    const Tensor &input_,
+    const Tensor &weight_,
+    IntArrayRef kernel_size,
+    const c10::optional<Tensor> &bias_opt,
+    IntArrayRef stride,
+    IntArrayRef padding,
+    IntArrayRef dilation,
+    const Tensor &out) {
+  TORCH_CHECK(kernel_size.size() == 2);
+  TORCH_CHECK(stride.size() == 2);
+  TORCH_CHECK(padding.size() == 2);
+  TORCH_CHECK(dilation.size() == 2);
+
+  auto input = input_.expect_contiguous();
+  auto weight = weight_.expect_contiguous();
+  auto bias = [&] {
+    if (bias_opt.has_value() && bias_opt->defined()) {
+      return bias_opt->expect_contiguous();
+    }
+    return c10::MaybeOwned<Tensor>::owned(c10::in_place);
+  }();
+
+  conv_depthwise2d_forward_out(
+      *input,
+      out,
+      *weight,
+      *bias,
+      kernel_size[1], kernel_size[0],
+      stride[1], stride[0],
+      padding[1], padding[0],
+      dilation[1], dilation[0]);
+  return out;
+}
+
+Tensor conv_depthwise2d_cuda(
+    const Tensor &input,
+    const Tensor &weight,
+    IntArrayRef kernel_size,
+    const c10::optional<Tensor> &bias,
+    IntArrayRef stride,
+    IntArrayRef padding,
+    IntArrayRef dilation) {
+  auto out = at::empty({0}, input.options());
+  return conv_depthwise2d_cuda_out(input, weight, kernel_size, bias,
+                                   stride, padding, dilation, out);
+}
+
+std::tuple<Tensor&, Tensor&> conv_depthwise2d_backward_cuda_out(
+    const Tensor & grad_output_,
+    const Tensor & self_,
+    const Tensor & weight_,
     IntArrayRef kernel_size,
     IntArrayRef stride,
     IntArrayRef padding,
     IntArrayRef dilation,
     Tensor & grad_input,
     Tensor & grad_weight) {
+  auto self = self_.expect_contiguous();
+  auto grad_output = grad_output_.expect_contiguous();
+
   if (grad_weight.defined()) {
-    grad_weight.resize_(weight.sizes());
-    grad_weight.zero_();
+    conv_depthwise2d_grad_weight_out(
+        *self, *grad_output, grad_weight,
+        kernel_size[1], kernel_size[0],
+        stride[1], stride[0],
+        padding[1], padding[0],
+        dilation[1], dilation[0]);
   }
 
-  return legacy::cuda::_thnn_conv_depthwise2d_backward_out(grad_input, grad_weight,
-                                                           grad_output, self, weight,
-                                                           kernel_size, stride, padding, dilation);
+  if (grad_input.defined()) {
+    auto weight = weight_.expect_contiguous();
+    conv_depthwise2d_backward_out(
+        *self, *grad_output, grad_input, *weight,
+        kernel_size[0], kernel_size[1],
```

Comment:
should it be `kernel_size[1], kernel_size[0]` here, like in the forward call? Arguments are `int kW, int kH,`. This legacy hw flip is so confusing